ks_patches/vaksine_utils.js

```diff
@@ -0,0 +1,170 @@
+            return {
+                name: "Comirnaty (BioNTech og Pfizer)",
+                type: "Comirnaty"
+            };
+        case "JAN03":
+            return {
+                name: "Covid-19 Vaccine Janssen",
+                type: "Janssen"
+            };
+        case "CUR03":  // Not supported in DHIS2 yet
+        default:
+            return {
+            name: vaccine.vaccineCode.display + ' (Annet)',
+                type: "annenvaksine"
+            };
+    }
+}
+
+
+function enrichWithValidation(vaccine) {
+    if(!vaccine.profileDose) {
+        return {...vaccine, vaccineIsInProfile: false, dateMismatch: false, nameMismatch: false, updatePossible: true};
+    }
+    var dateMismatch = vaccine.date !== vaccine.profileDose.date;
+    var nameMismatch = vaccine.name !== vaccine.profileDose.name && vaccine.type !== vaccine.profileDose.name;
+    return {...vaccine,
+        vaccineIsInProfile: true,
+        dateMismatch,
+        nameMismatch,
+        updatePossible:  dateMismatch || nameMismatch
+    }
+}
+
+export function saveVaccineToProfile(tei, vaccines, attributesById, TEIService, $q) {
+
+    var teiCopy = angular.copy(tei);
+
+    teiCopy.attributes = getUpdatedVaccineAttributes(teiCopy.attributes, vaccines);
+    if(!tei.trackedEntityInstance) { // If trackedEntityInstance is not set, we assume we are registering
+        updateAttributes(tei, teiCopy.attributes);
+        var promiseMaker = $q.defer();
+        promiseMaker.resolve(true);
+        return promiseMaker.promise;
+    }
+    return TEIService.update(teiCopy, [], attributesById).then((status) => {
+        if(status && status.httpStatus === "OK") {
+            updateAttributes(tei, teiCopy.attributes);
+            return true;
+        }
+        return false;
+    });
+}
+
+function getUpdatedVaccineAttributes(attributes, vaccines) {
+    return [
+        getUpdatedAttribute(attributes, vaccines[0].type, PROFIL_VAKSINE_1_TYPE_ID),
+        getUpdatedAttribute(attributes, vaccines[0].date, PROFIL_VAKSINE_1_DATO_ID),
+        getUpdatedAttribute(attributes, vaccines[1].type, PROFIL_VAKSINE_2_TYPE_ID),
+        getUpdatedAttribute(attributes, vaccines[1].date, PROFIL_VAKSINE_2_DATO_ID),
+        getUpdatedAttribute(attributes, vaccines[2].type, PROFIL_VAKSINE_3_TYPE_ID),
+        getUpdatedAttribute(attributes, vaccines[2].date, PROFIL_VAKSINE_3_DATO_ID),
+    ];
+}
+
+function getUpdatedAttribute(attributes, value, id) {
+    var attribute = attributes.find(att => att.attribute === id);
+    return {...attribute, value};
+}
+
+function updateAttributes(tei, attributesToUpdate) {
+    tei.attributes = tei.attributes.map(att => getUpdatedAttributeOrSelf(att, attributesToUpdate))
+}
+
+function getUpdatedAttributeOrSelf(attribute, attributesToUpdate) {
+    return attributesToUpdate.find(att => att.attribute === attribute.attribute) || attribute;
+}
+
+export function hackToUpdateVaccineFieldsInProfile(selectedTei, sysvakVaccines) {
```

Comment:
What would be the non-hacky way of doing this? It seems that we are just maing a mapping here, that perhaps can live as a constant somewhere?

Reply:
The problem is the strange way attribute values some times are stored directly on the tei rather than on the attribute. Tried in vain for ages to just update the correct attribute in the attribute array, but since there are these attributes on the tei directly, which apparently are being used for (some of the?) values in the profile, we need to update both places. 
Not sure how to avoid this.